### DISCUSSION TITLE
useAsyncData vs useFetch

### DISCUSSION THREAD

**Discussion by vencho-mdp** [nuxt/nuxt#16063] [Ideas]
As with `asyncData` and `fetch`, these two composables are not well differentiated in the docs. When should we use one or another? When should we `useFetch` in pages? When is it bad to fetch the data with `asyncData` and pass it as props?

**Comment by vencho-mdp** [comment]
@danielroe Thoughts about this?

**Reply by danielroe** [reply]
How about this? https://v3.nuxtjs.org/docs/migration/component-options#using-composables

**Reply by vencho-mdp** [reply]
It helps, but IMO it isn't clear enough. Maybe adding some use cases about when to use one or another could help. For example: what do you mean by a "simple fetch"  in the comment in the code section?

**Reply by danielroe** [reply]
useFetch receives a URL and gets that data.

useAsyncData might have more complex logic.

`useFetch(url)` is exactly equivalent to `useAsyncData(url, () => $fetch(url))` - it's DX for the most common use case.

**Reply by mklueh** [reply]
@danielroe  And in what case to we have to bother with useAsyncData then? and what impact does the key have?

it says: "key: a unique key to ensure that data fetching can be properly de-duplicated across requests" - is this for caching?

**Comment by stafyniaksacha** [comment]
`useAsyncData` allows you to wait for an async function call to be resolved (can be a graphql request, custom axios call, etc...) while `useFetch` is a simple wrapper for the [Web Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)

**Reply by vencho-mdp** [reply]
Uh? Daniel said that `useFetch` is just a wrapper around `useAsyncData`.  Nuxt uses `h3` under the hood for data fetching.

**Reply by stafyniaksacha** [reply]
yep, `useFetch(url)` => `useAsyncData(url, () => $fetch(url))`

and the `$fetch` is from https://github.com/unjs/ohmyfetch

**Reply by mklueh** [reply]
So **useFetch** is a wrapper around **useAsyncData**, which uses **$fetch**?

**Comment by mklueh** [comment]
This is the most unclear thing about Nuxt and I find myself googling it every time. There are also so many YouTube videos about that topic, it confuses everyone who is new to Nuxt https://www.youtube.com/results?search_query=nuxt+fetch+vs+asyncdata

There should be clear examples on when to use what at the top of the doc page and in my opinion, fetch should be mentioned first and the other one second as a "special case"

**Reply by kissu** [reply]
This one explains it somewhat: https://stackoverflow.com/questions/49251437/difference-between-asyncdata-vs-fetch

**Reply by vencho-mdp** [reply]
Actually, that doesn't help since `asyncData` and `fetch` are completely different in Nuxt 2 and Nuxt 3 (though they almost have the same name)

**Comment by muellercornelius** [comment]
I am sorry to extend this topic even further. I understand the difference between useFetch and useAsyncData... but why do we have $fetch when we have useFetch and vice versa?

**Reply by vencho-mdp** [reply]
AFAIK `$fetch` is the actual method to fetch data from the `ohmyfetch` package. Indeed, when you're using `useFetch` it uses it under the hood. You could use another lib, but that's the preferred way for a couple of reasons, like direct function invocation in server when fetching data instead of an API call.

**Comment by memic84** [comment]
I must admit that although i understand that `useFetch` is a wrapper around `useAsyncData`, and also that $fetch (h3) is used in both of them, that it might be confusing not only for newcomers but also people migrating from Nuxt 2, where `fetch` and `asyncData` methods were a bit different? I do have a couple of questions:

- `useFetch` and `useAsyncData` are both executed on the server in SSR mode (on first load), on page navigation they are executed from the client? Is this correct?  Can we see the `useAsyncData` as a lifecycle method?
- Do `useFetch` and `useAsyncData` work outside setup function? In composables, components, services?
- When is it better to just use $fetch? Is $fetch always going to be executed on the client, unless used in asycData?

For now as i understand it, it's `useFetch` as default, and when having more complex logic `useAsyncData`.

**Comment by hymair** [comment]
I feel like this is an overdone abstraction. At some point abstractions become too obfuscating and end up confusing the users (devs) instead of helping. As opinionated as Nuxt is a single unified data fetching method is more than enough.
